### PR TITLE
add nixos sandbox utilities, fixes #16

### LINKED ISSUE #16: Utilities for operating with nix-shell environments
Hi,

independently from this repository, I developed a few utility functions for working with nix-shell environments [link](https://github.com/svenkeidel/emacs.d/blob/master/nixos.el). I think these functions could be beneficial for this repository and I would like to merge my work with yours.

The functions include:
* `nix-shell-command` to compose a nix-shell command string that can be passed to the commandline
* `nix-shell` that directly executes a command in the given sandbox
* `nixos-find-sandbox` searches from the given path upwards until it finds a `*.nix` file.
* `nixos-current-sandbox` searches from the current working directroy upwards until it finds a `*.nix` file
* `nixos-executable-find`, a replacement for the built-in `executable-find`. It searches in the given sandbox for executables.

Let me know what you think.

Best,
Sven

### PR DESCRIPTION
**Pull request by svenkeidel**
So, I stripped out the Haskell related stuff, added some autoload annotations and a copyright and license notice. I'm open for suggestions and improvements.

### CODE REVIEW

File: nixos-sandbox.el [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+;;; Commentary:
+
+;; Useful functions for working with nix-shell sandboxes
+
+;;; Code:
+
+(require 'dash)
+
+(defgroup nixos nil
```
Comment (CestDiego) [inline nixos-sandbox.el]: This is not really nixos specific, is it? we could have a group `nix`

**Comment by travisbhartwell** [comment]
Looks good, sorry I'm just now getting to it.

As @CestDiego commented inline, these functions aren't necessarily NixOS-specific, they could be used on another system that just has the Nix package manager installed.  It still belongs in our repo, but I'd probably do a `s/nixos/nix/` in your changes.

Given those changes, I'm not opposed to merging this.  Afterwards, we should do a PR on MELPA so this can be included as a package in its own right.

**Comment by CestDiego** [comment]
+

**Comment by svenkeidel** [comment]
So, finally I have the time to do some open source work. I made the changes you requested, fixed a few bugs and made the shell command more reliable. The code is [here](https://github.com/svenkeidel/nix-emacs/tree/nix-sandbox).

Best,
Sven

**Comment by CestDiego** [comment]
I approve, this looks very cool. It would be cool if you also included in the PR the documentation you have in your branch :)

**Comment by wedens** [comment]
I've tried to use it (just copy-pasted contents to my `.emacs`) and flycheck is not reporting any errors

**Comment by CestDiego** [comment]
I've tried this and this looks good. @wedens thank you for the report.

**Comment by CestDiego** [comment]
@svenkeidel thank you, good job. You can safely delete your branch

**Comment by svenkeidel** [comment]
@wedens, [here](https://github.com/svenkeidel/emacs.d/blob/master/config.org#flycheck) is the necessary glue code to let flycheck now about your sandbox environments.

**Comment by wedens** [comment]
@svenkeidel I've been trying to use it with exaclty this glue code :/

**Comment by Pitometsu** [comment]
Awesome enhancement!

But if I, for example, would use nix sandbox for Rails development, then I should add `nix-mode` wrappers as a number of separate PR to each of:
+ robe-mode
+ enh-ruby-mode
+ rspec-mode
+ rdoc-mode
+ rubocop
+ projectile-rails
+ rails-new
+ pry
+ bundler
+ inf-ruby
+ ac-inf-ruby
+ etc...

Not so cool :(

**Comment by Pitometsu** [comment]
Maybe better just overwrite some `shell-` commands?

**Comment by svenkeidel** [comment]
@wedens, Please open a new issue and send me your `default.nix`\`shell.nix` file and the relevant section of your emacs configuration.

**Comment by svenkeidel** [comment]
@Pitometsu, It is not that simple to just activate `nix-mode` for all plugins you mentioned. The plugins have to provide hooks and then it might be possible to make the nix aware.

**Pull request merged**